dataplane/forwarding/fwd.go

```diff
@@ -124,54 +192,111 @@ func (f *Engine) UpdatePacketSink(contextID *fwdpb.ContextId, packet fwdcontext.
 	return nil
 }
 
-// ContextCreate creates a new context.
-func (f *Engine) ContextCreate(contextID *fwdpb.ContextId) (err error) {
-	defer func() {
-		if err != nil {
-			log.Errorf("%v: %v, ContextCreate failed", f.name, err)
+// ContextCreate creates a new context. Note that if the packet sink and/or
+// notification services are specified but not reachable, the context creation
+// fails.
+func (e *Engine) ContextCreate(_ context.Context, request *fwdpb.ContextCreateRequest) (*fwdpb.ContextCreateReply, error) {
+	paddr := request.GetPacketAddress()
+	naddr := request.GetNotificationAddress()
+	ns, err := e.GetNotificationCallback(naddr)
+	if err != nil {
+		return nil, err
+	}
+
+	ps, err := e.GetPacketSinkCallback(paddr)
+	if err != nil {
+		return nil, err
+	}
+
+	cid := request.GetContextId()
+	if err := e.contextCreateByID(cid); err != nil {
+		return nil, err
+	}
+
+	if err := e.UpdateNotification(cid, ns, naddr); err != nil {
+		return nil, err
+	}
+	return &fwdpb.ContextCreateReply{}, e.UpdatePacketSink(cid, ps, paddr)
+}
+
+// ContextUpdate updates a forwarding context. Note that if the packet sink and/or
+// notification services are specified but not reachable, the context update
+// fails.
+func (e *Engine) ContextUpdate(_ context.Context, request *fwdpb.ContextUpdateRequest) (*fwdpb.ContextUpdateReply, error) {
+	var err error
+	var ps fwdcontext.PacketCallback
+	var ns fwdcontext.NotificationCallback
+
+	paddr := request.GetPacketAddress()
+	naddr := request.GetNotificationAddress()
+	for _, op := range request.GetOperations() {
+		switch op {
+		case fwdpb.ContextUpdateRequest_OPERATION_UPDATE_PACKET_ADDRESS:
+			if ps, err = e.GetPacketSinkCallback(paddr); err != nil {
+				return nil, err
+			}
+
+		case fwdpb.ContextUpdateRequest_OPERATION_UPDATE_NOTIFICATION_ADDRESS:
+			if ns, err = e.GetNotificationCallback(naddr); err != nil {
+				return nil, err
+			}
+
 		}
-	}()
+	}
+
+	cid := request.GetContextId()
+	for _, op := range request.GetOperations() {
+		switch op {
+		case fwdpb.ContextUpdateRequest_OPERATION_UPDATE_PACKET_ADDRESS:
+			if err = e.UpdatePacketSink(cid, ps, paddr); err != nil {
+				return nil, err
+			}
+
+		case fwdpb.ContextUpdateRequest_OPERATION_UPDATE_NOTIFICATION_ADDRESS:
+			if err = e.UpdateNotification(cid, ns, naddr); err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
Should the second for loop just be merged with the first? It seems to be calling update multiple times so it's confusing what the intention of the first for loop is.

Reply:
good catch